src/_pytest/recwarn.py

```diff
@@ -286,11 +290,13 @@ def __exit__(
                                 break
                     else:
                         fail(
-                            "DID NOT WARN. No warnings of type {} matching"
-                            " ('{}') was emitted. The list of emitted warnings"
-                            " is: {}.".format(
-                                self.expected_warning,
-                                self.match_expr,
-                                [each.message for each in self],
+                            "DID NOT WARN. No warnings of type {expected} matching the"
+                            " regex was emitted.\n"
+                            "The regex is: {match!r}\n"
+                            "The list of emitted warnings"
+                            " is: [{found}].".format(
```

Comment:
Would join the two lines just to make the message easier to grep for.

```suggestion
                            "The list of emitted warnings is: [{found}].".format(
```